scRNA-seq-advanced/03-dataset_integration.Rmd

```diff
@@ -0,0 +1,550 @@
+A well-integrated dataset will instead show much more mixing among batches, as the batch effects will have hopefully been corrected.
+How might you expect an integrated UMAP to look?
+
+
+
+
+### Integration with `fastMNN`
+
+Finally, we're ready to integrate!
+To start, we'll use the `fastMNN` approach from the Bioconductor [`batchelor` package](http://www.bioconductor.org/packages/3.16/bioc/html/batchelor.html).
+As the name suggests, this method is pretty fast! 
+It returns a new SCE object containing integrated results:
+
+```{r}
+integrated_sce <- batchelor::fastMNN(
+  # the merged SCE object
+  merged_sce,
+  # vector indicating the batches, which we created as `$sample` when formatting SCEs
+  batch = merged_sce$sample, 
+  # Optionally, we can specify genes to use as the previously-ID'd
+  #  high variance genes
+  subset.row = hv_list
+)
+
+# Let's have a look!
+integrated_sce
+```
+
+There are couple pieces of information here of interest:
+
+- The new `reconstructed` assay represents the batch-corrected normalized expression values.
+While these are calculated and used during integration, they are **NOT** stand-alone values that you should use for other applications, e.g. differential gene expression.
+- The new `corrected` reduced dimension represents the principle components (PCs) of these corrected expression values. 
+
+
+To conveniently keep everything in one place, let's add the PCs information (with an informative name!) into our `merged_sce` object:
+
+```{r fastmnn pcs}
+# Make a new reducedDim named fastMNN_PCA from the corrected reducedDim in integrated_sce
+reducedDim(merged_sce, "fastMNN_PCA") <- reducedDim(integrated_sce, "corrected")
+
+merged_sce
+```
+
+Finally, we'll calculate UMAPs from these corrected PCs and plot them so we can see how qualitatively successful integration was:
+
+```{r fastmnn umap}
+# Calculate UMAP
+merged_sce <- scater::runUMAP(
+  merged_sce, 
+  # create UMAP from these PCs:
+  dimred = "fastMNN_PCA", 
+  # name the UMAP:
+  name = "fastMNN_UMAP"
+)
+```
+
+First, let's plot the integrated UMAP highlighting the different batches.
+A well-integrated dataset will show batch mixing, but a poorly-integrated dataset will show more separation among batches, similar to the uncorrected UMAP.
+
+```{r fastmnn umap batches}
+scater::plotReducedDim(merged_sce,
+                       # plot the fastMNN coordinates
+                       dimred = "fastMNN_UMAP",
+                       # color by sample (batch)
+                       colour_by = "sample",
+                       # Some styling to help us see the points:
+                       point_size = 0.5,
+                       point_alpha = 0.2) +
+  ggplot2::ggtitle("UMAP after integration with fastMNN") +
+  # Modify the ggplot2 guides so that the legend points are larger and easier to see:
+  ggplot2::guides(colour = ggplot2::guide_legend(override.aes = list(size = 3, alpha = 1)))
+```
```

Comment:
I think this (or the next plot) might be a good place to show faceting. Randomizing the order to show hidden points is going to probably be more difficult, but faceting can provide the same info.